client.go

```diff
@@ -48,17 +51,45 @@ type Client struct {
 }
 
 func New(writeKey string) *Client {
+	return NewWithConfig(writeKey, Config{})
+}
+
+func NewWithConfig(writeKey string, config Config) *Client {
+	conf := getFinalConfig(config)
 	return &Client{
-		BaseEndpoint:     DefaultBaseEndpoint,
-		Logger:           log.New(os.Stderr, "segment ", log.LstdFlags),
-		writeKey:         writeKey,
-		Client:           http.DefaultClient,
-		cmap:             newConcurrentMap(),
-		MaxBatchBytes:    500 << 10,
-		MaxBatchCount:    100,
-		MaxBatchInterval: 10 * time.Second,
-		semaphore:        make(semaphore.Semaphore, 10),
+		Config:    conf,
+		writeKey:  writeKey,
+		cmap:      newConcurrentMap(),
+		semaphore: make(semaphore.Semaphore, 10),
+	}
+}
+
+func getFinalConfig(c Config) Config {
```

Comment:
`getFinalConfig` doesn't really explain much.
Could you consider rename/refactor to 
```
func applyDefaultConfig(c *Config){
// sets default inline
}
```

Reply:
naming is hard. How do parents do this